File: src/util/config/types/SecurityConfiguration.ts
/*
	Spacebar: A FOSS re-implementation and extension of the Discord.com backend.
	Copyright (C) 2023 Spacebar and Spacebar Contributors
	
	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU Affero General Public License as published
	by the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.
	
	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU Affero General Public License for more details.
	
	You should have received a copy of the GNU Affero General Public License
	along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import crypto from "crypto";
import { CaptchaConfiguration, TwoFactorConfiguration } from ".";

export class SecurityConfiguration {
	captcha: CaptchaConfiguration = new CaptchaConfiguration();
	twoFactor: TwoFactorConfiguration = new TwoFactorConfiguration();
	autoUpdate: boolean | number = true;
	requestSignature: string = crypto.randomBytes(32).toString("base64");
	jwtSecret: string = crypto.randomBytes(256).toString("base64");
	// header to get the real user ip address
	// X-Forwarded-For for nginx/reverse proxies
	// CF-Connecting-IP for cloudflare
	forwardedFor: string | null = null;
	ipdataApiKey: string | null = "eca677b284b3bac29eb72f5e496aa9047f26543605efe99ff2ce35c9";
	mfaBackupCodeCount: number = 10;
	statsWorldReadable: boolean = true;
	defaultRegistrationTokenExpiration: number = 1000 * 60 * 60 * 24 * 7; //1 week
}

File: extra/admin-api/Spacebar.ConfigModel/SecurityConfiguration.cs
using System.Text.Json.Serialization;

namespace Spacebar.ConfigModel;

public class SecurityConfiguration {
    [JsonPropertyName("captcha")]
    public CaptchaConfiguration Captcha { get; set; } = new();

    [JsonPropertyName("twoFactor")]
    public TwoFactorConfiguration TwoFactor { get; set; } = new();

    [JsonPropertyName("autoUpdate")]
    public bool AutoUpdate { get; set; } = true;

    [JsonPropertyName("requestSignature")] public string RequestSignature; // {get;set;}=crypto.randomBytes(32).toString("base64");

    [JsonPropertyName("jwtSecret")]
    public string? JwtSecret { get; set; } = null;

    [JsonPropertyName("forwardedFor")]
    public string? ForwardedFor { get; set; } = null;

    [JsonPropertyName("trustedProxies")]
    public string TrustedProxies { get; set; } = null;

    [JsonPropertyName("abuseIpDbApiKey")]
    public string? AbuseIpDbApiKey { get; set; } = null;

    [JsonPropertyName("abuseipdbBlacklistRatelimit")]
    public int AbuseipdbBlacklistRatelimit { get; set; } = 5;

    [JsonPropertyName("abuseipdbConfidenceScoreTreshold")]
    public int AbuseipdbConfidenceScoreTreshold { get; set; } = 50;

    [JsonPropertyName("ipdataApiKey")]
    public string? IpdataApiKey { get; set; } = null;

    [JsonPropertyName("mfaBackupCodeCount")]
    public int MfaBackupCodeCount { get; set; } = 10;

    [JsonPropertyName("statsWorldReadable")]
    public bool StatsWorldReadable { get; set; } = true;

    [JsonPropertyName("defaultRegistrationTokenExpiration")]
    public int DefaultRegistrationTokenExpiration { get; set; } = 1000 * 60 * 60 * 24 * 7;

    [JsonPropertyName("cdnSignUrls")]
    public bool CdnSignUrls { get; set; } = false;

    [JsonPropertyName("cdnSignatureKey")]
    public string CdnSignatureKey { get; set; } = null!; // crypto.randomBytes(32).toString("base64");

    [JsonPropertyName("cdnSignatureDuration")]
    public string CdnSignatureDuration { get; set; } = "24h";

    [JsonPropertyName("cdnSignatureIncludeIp")]
    public bool CdnSignatureIncludeIp { get; set; } = true;

    [JsonPropertyName("cdnSignatureIncludeUserAgent")]
    public bool CdnSignatureIncludeUserAgent { get; set; } = true;
}

public class CaptchaConfiguration {
    [JsonPropertyName("enabled")]
    public bool Enabled { get; set; } = false;

    /// <summary>
    /// One of: null, recaptcha, hcaptcha
    /// </summary>
    [JsonPropertyName("service")]
    public string Service { get; set; } = "none";

    [JsonPropertyName("sitekey")]
    public string? SiteKey { get; set; } = null;

    [JsonPropertyName("secret")]
    public string? Secret { get; set; } = null;
}

public class TwoFactorConfiguration {
    [JsonPropertyName("generateBackupCodes")]
    public bool GenerateBackupCodes { get; set; } = true;
}